<a href="https://colab.research.google.com/github/YouCanKeepSilence/ml_labs/blob/master/ML_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Импорт данных

In [0]:
import numpy as np
import pandas as pd

In [10]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Загружаем данные из .csv

In [0]:
names_array = ['gender', 'age', 'first_label', 'second_label', 'third_label', *['c_{}'.format(i + 1) for i in range(60000)]]
url = '/content/drive/My Drive/ML/half.csv'
data = pd.read_csv(url, names=names_array)

# Обработка данных

Удаляем ненужные стобцы (по причине малого количества данных в них)

In [0]:
data = data.drop(['second_label', 'third_label'], axis=1)

In [13]:
#new_data = data # to skip augmentation because it'll break PCA method
#new_batch = 5000
#sensors_count = 12
data

gender  age  first_label  ...   c_59998   c_59999   c_60000
0     Female   85            3  ... -0.042003 -0.034466 -0.039000
1       Male   79            9  ... -0.819597 -0.771195 -0.845490
2       Male   66            3  ...  0.030947  0.034947  0.037947
3       Male   93            2  ... -0.016540 -0.016289 -0.022542
4       Male   65            2  ... -0.026058 -0.012157 -0.001676
...      ...  ...          ...  ...       ...       ...       ...
1245    Male   74            2  ...  0.675640  0.669640  0.661640
1246  Female   85            7  ... -0.053092 -0.049083 -0.049954
1247    Male   61            2  ...  0.072895  0.062235  0.060652
1248  Female   85            3  ... -0.032233 -0.030831 -0.033880
1249    Male   82            1  ... -0.035582 -0.215320 -0.406685

[1250 rows x 60003 columns]

## Агументация данных

In [0]:
# batch_size = 1000
new_batch = 250
measurments_for_sensors = 5000
sensors_count = 12

def augmentate():
  measurment_names = data.columns[3:]
  new_data_inner = pd.DataFrame()
  # Пользуемся тем, что имеем 5к измерений для каждого из 12 датчиков. Соответственно нарежем по 250 с каждого из 12 датчиков
  for i in range(0, measurments_for_sensors, new_batch):
      current_measurment_names = {f'c_{k + j * measurments_for_sensors + 1}': f'c_{(new_batch * j + k % new_batch) % (new_batch * sensors_count)  + 1}' for j in range(sensors_count) for k in range(i, i + new_batch)}
      mapped_piece = data[['age', 'gender', 'first_label', *[f'c_{k + j * measurments_for_sensors + 1}'  for j in range(sensors_count) for k in range(i, i + new_batch)]]].rename(columns=current_measurment_names)
      new_data_inner = pd.concat([new_data_inner, mapped_piece])
  return new_data_inner

# don't call augmentate...
new_data = data

Таким образом получаем из размерности 1250, 60003 -> 25000, 3003 увеличив количество записей в датасете в 20 раз



## Обработаем пропущенные значения

In [15]:
print(new_data['age'].unique())
print(new_data['gender'].unique())
print(new_data['first_label'].unique())
print(new_data.isnull().sum())

[85 79 66 93 65 55 82 87 51 61 90 77 70 60 62 59 84 71 45 48 75 49 72 78
 86 46 94 33 74 64 67 63 88 32 92 68 76 54 89 31 80 30 57 83 34 27 73 69
 52 44 43 47 39 56 81 53 50 16 29 37 58 24 22 40 41 11 26 28 96 35 91 25
 36 23 19 17 42 18 15  8 20 38 14 21]
['Female' 'Male']
[3 9 2 5 1 7 6 4 8]
gender         0
age            0
first_label    0
c_1            0
c_2            0
              ..
c_59996        0
c_59997        0
c_59998        0
c_59999        0
c_60000        0
Length: 60003, dtype: int64


Пропущенных значений нет.

## Обработаем категориальные признаки

In [18]:
categorial_columns = ['gender', 'first_label']
numerical_columns = ['age', *[f'c_{i + 1}' for i in range(new_batch * sensors_count)]]
print(categorial_columns)

['gender', 'first_label']


Проведем бинаризацию пола, label не трогаем, им займемся дальше

In [19]:
new_data.at[data['gender'] == 'Male', 'gender'] = 0
new_data.at[data['gender'] == 'Female', 'gender'] = 1
new_data['gender'].describe()

count    1250.000000
mean        0.392800
std         0.488568
min         0.000000
25%         0.000000
50%         0.000000
75%         1.000000
max         1.000000
Name: gender, dtype: float64

In [21]:
categorial_columns.remove('gender')
numerical_columns.append('gender')

ValueError: ignored

## Произведем нормировку данных

In [22]:
numerical_data = new_data[numerical_columns]
numerical_data.describe()

age          c_1  ...       c_3000       gender
count  1250.000000  1250.000000  ...  1250.000000  1250.000000
mean     65.856000    -0.001139  ...    -0.001172     0.392800
std      16.259704     0.109153  ...     0.148407     0.488568
min       8.000000    -1.016408  ...    -1.517420     0.000000
25%      58.000000    -0.033989  ...    -0.045384     0.000000
50%      68.000000    -0.011761  ...    -0.016034     0.000000
75%      78.000000     0.017391  ...     0.014974     1.000000
max      96.000000     1.042934  ...     1.326323     1.000000

[8 rows x 3002 columns]

In [0]:
mean = numerical_data.mean(axis = 0)
std = numerical_data.std(axis = 0)
numerical_data = (numerical_data - mean) / std
new_data[numerical_columns] = numerical_data

In [24]:
new_data.describe()

gender           age  ...      c_59999      c_60000
count  1.250000e+03  1.250000e+03  ...  1250.000000  1250.000000
mean  -4.112266e-17  3.527845e-16  ...    -0.012375    -0.014290
std    1.000000e+00  1.000000e+00  ...     0.260574     0.281653
min   -8.039815e-01 -3.558244e+00  ...    -3.813070    -4.095789
25%   -8.039815e-01 -4.831576e-01  ...    -0.081939    -0.085807
50%   -8.039815e-01  1.318597e-01  ...    -0.032070    -0.034636
75%    1.242815e+00  7.468771e-01  ...     0.021461     0.023269
max    1.242815e+00  1.853908e+00  ...     1.846399     2.268871

[8 rows x 60003 columns]

Нормализовали все поля кроме label т.к. он является выходным и его нет смысла нормализовывать.

# Подготовка данных для модели

In [25]:
from sklearn.model_selection import train_test_split
y = new_data['first_label'].values
X = new_data.drop(['first_label'], axis=1).values

x_train, x_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 42)

n_train, _ = x_train.shape 
n_test,  _ = x_test.shape 

print('Train dataset size: {}'.format(n_train))
print('Test dataset size: {}'.format(n_test))

Train dataset size: 875
Test dataset size: 375


# Создание и обучение моделей

## KMeans

In [0]:
from sklearn.cluster import KMeans
kmeans = KMeans(n_clusters=len(new_data['first_label'].unique()), random_state=42).fit(X)

In [27]:
kmeans_cluster = kmeans.predict(X) + 1
print(np.unique(kmeans.labels_))
check = kmeans_cluster == y
np.where(check == True)[0].shape

[0 1 2 3 4 5 6 7 8]


(90,)

Т.к. разбиение подразумевает передачу количества классов, то количество классов совпадает, однако пересечение с реальным разбиением достаточно слабое. (90)

## DBSCAN

In [0]:
from sklearn.cluster import DBSCAN
dbscan = DBSCAN().fit(X)

In [29]:
np.unique(dbscan.labels_)

array([-1])

На агументированных данных с таким набором параметров разбиение происходит на 5 классов, в то время как изначальное разбиение произведено на 9 классов.
На неагументированных данных не обучается. (вернуло массив, содержащий только -1)

## Affinity Propagation


In [30]:
from sklearn.cluster import AffinityPropagation
affinity = AffinityPropagation().fit(X)

/usr/local/lib/python3.6/dist-packages/sklearn/cluster/_affinity_propagation.py:226: ConvergenceWarning: Affinity propagation did not converge, this model will not have any cluster centers.
  "will not have any cluster centers.", ConvergenceWarning)


In [31]:
np.unique(affinity.labels_)

array([-1])

На аугментированных данных - не удалось проверить ибо вылезло за пределы по памяти (более 25 Гб оперативки).
На неаугментированных - не обучилось. (вернуло массив, содержащий только -1)

## Spectral Clustering


In [36]:
from sklearn.cluster import SpectralClustering
spectral = SpectralClustering(n_clusters=len(new_data['first_label'].unique()), random_state=42).fit_predict(X)

/usr/local/lib/python3.6/dist-packages/sklearn/manifold/_spectral_embedding.py:236: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.
  warnings.warn("Graph is not fully connected, spectral embedding"


In [38]:
# spectral_cluster = spectral.predict(X) +
check_spectral = spectral + 1 == y
# check_spectral = spectral_cluster == y
np.where(check_spectral == True)[0].shape

(169,)

На неаугментированных данных разбиение имеет большее пересечение с оригинальным, чем KMeans. (169)
На аугментированных обучение продолжается очень долго (более двух часов) не дождался.

## Birch


In [0]:
from sklearn.cluster import Birch
birch = Birch(n_clusters=len(new_data['first_label'].unique())).fit(X)

In [39]:
birch_cluster = birch.predict(X) + 1
print(np.unique(birch.labels_))
check_birch = birch_cluster == y
np.where(check_birch == True)[0].shape

[0 1 2 3 4 5 6 7 8]


(169,)

На неагументированных данных имеет такое же количество пересечений с оригинальным разбиением как и Spectral Clustering. (169)

# Вывод
Были опробованы следующие методы кластеризации: KMeans, DBscan, Affinity Propagation, Spectral Clustering, Birch.

Подвердить или опровергнуть достоверность их разбиения непредставляется возможным. Однако, наибольшее пересечение с оригинальным разбиением дали Birch и Spectral Clustering